### PR TITLE
Add comprehensive multiplayer event logging with game seed

### PR DESCRIPTION
**Pull request by skypher**
## Summary
- Add comprehensive multiplayer event logging with game seed tracking:
  - Log game creation and joins with game name, privacy level, connection type, player ID, and full game seed
  - Log player joins/leaves with connection type, player count, and detailed leave reasons  
  - Log game exit with reason description
- Convert leave reasons from magic numbers to type-safe `leaveinfo_t` enum (moved to `dvlnet/leaveinfo.hpp` to avoid libsodium header dependencies in non-network code)
- Add `multi_logging_test.cpp` with test coverage for leave reason formatting edge cases
- Improve packet parsing safety with better error handling and validation in delta import and command parsing
- Guard libsodium-specific code with `#ifdef PACKET_ENCRYPTION` where needed

## Testing
- Added unit tests for leave reason formatting (`multi_logging_test.cpp`)
- Existing multiplayer functionality preserved
- Relying on CI for cross-platform verification

### CODE REVIEW

File: Source/multi.cpp [hunk 1]
```diff
@@ -520,6 +530,26 @@ bool InitMulti(GameData *gameData)
...
+{
+	switch (leaveReason) {
+	case LEAVE_EXIT:
+		return "normal exit";
+	case LEAVE_ENDING:
+		return "Diablo defeated";
+	case LEAVE_DROP:
+		return "connection timeout";
+	default:
+		return fmt::format("code 0x{:08X}", leaveReason);
```
Comment (AJenbo) [inline Source/multi.cpp]: Can we get rid of this by making the leave reasons in to a finite enum?

**Comment by skypher** [comment]
Updated. Still need to do a manual run test.

**Comment by skypher** [comment]
LGTM

**Comment by AJenbo** [comment]
Looks like there has been some other changes to one of the files, could you please resolve the conflict.

**Comment by skypher** [comment]
> Looks like there has been some other changes to one of the files, could you please resolve the conflict.

rebased

**Comment by AJenbo** [comment]
Everything failed

**Comment by skypher** [comment]
> Everything failed

Dang, let me check again and make it right.

**Comment by skypher** [comment]
> > Everything failed
> 
> Dang, let me check again and make it right.

Apologies for the churn, it should work now.

**Comment by AJenbo** [comment]
No worries, you probably need `#ifdef PACKET_ENCRYPTION` to fix the last ones

**Comment by skypher** [comment]
> No worries, you probably need `#ifdef PACKET_ENCRYPTION` to fix the last ones

Thanks for the hint, will take a look

**Comment by skypher** [comment]
> > No worries, you probably need `#ifdef PACKET_ENCRYPTION` to fix the last ones
> 
> Thanks for the hint, will take a look

Alright, I think I've beaten them all into submission. RIP to anyone's inbox who received the build run notifs.

Had to make a bunch of platform config changes, be sure to take a look. I've documented the what and why in detail in the PR desc.

**Comment by StephenCWills** [comment]
> 4\. Several platforms (Android, iOS, 3DS, Switch, Vita, UWP) had `PACKET_ENCRYPTION=ON` by default but **no libsodium installed**. This misconfiguration went unnoticed because `packet.h` was previously only included in networking-specific files that might not compile on all platforms.

That can't be right. Android and 3DS definitely had libsodium. The other platforms in that list probably did too. Perhaps some CMake dependencies need to be adjusted instead.

**Comment by Trihedraf** [comment]
Yeah that's not right. All of those have libsodium, otherwise they wouldn't be able to join passworded tcp games, which they can.

**Comment by StephenCWills** [comment]
Probably the easiest way to fix this without having to adjust CMake dependencies would be to separate `leaveinfo_t` into its own header called `leaveinfo.hpp` and reference that in places that don't need the entirety of `packet.h`. This is probably a more correct solution anyway to avoid adding libsodium dependencies in modules that don't need it.

**Comment by skypher** [comment]
> > 4. Several platforms (Android, iOS, 3DS, Switch, Vita, UWP) had `PACKET_ENCRYPTION=ON` by default but **no libsodium installed**. This misconfiguration went unnoticed because `packet.h` was previously only included in networking-specific files that might not compile on all platforms.
> 
> That can't be right. Android and 3DS definitely had libsodium. The other platforms in that list probably did too. Perhaps some CMake dependencies need to be adjusted instead.

Good to know, I'll look into it.

**Comment by skypher** [comment]
> Probably the easiest way to fix this without having to adjust CMake dependencies would be to separate `leaveinfo_t` into its own header called `leaveinfo.hpp` and reference that in places that don't need the entirety of `packet.h`. This is probably a more correct solution anyway to avoid adding libsodium dependencies in modules that don't need it.

I can implement this, but I can't shake the feeling that something is off in the build configs and that it would be the right thing to fix this before it trips up anyone else in the future.

**Comment by AJenbo** [comment]
Our build system devides the source up into modules with dependencies. Only the net code should depend on libsodium, so it's headers are not being exposed to other parts of the code, Meaning that if you include a header from the network module (which depend on libsodium) in a game module, but do not declare the dependency then the sodium headers will be unavailable during compilation of the game module.

The network code is still part of the general libdevilutoinx module, which is what libsodium is being linked against: 

https://github.com/diasurgical/DevilutionX/blob/df325f6366f45dc4e37ed3bc355a45cd9d51a52d/Source/CMakeLists.txt#L1003

So when you used the packet.h header in the libdevilutionx_multiplayer module without changing it's dependencies you ended up get an error because that module isn't linked up against libsodium.

https://github.com/diasurgical/DevilutionX/blob/df325f6366f45dc4e37ed3bc355a45cd9d51a52d/Source/CMakeLists.txt#L591

So there isn't actually an issue here, just a complexity.

**Comment by StephenCWills** [comment]
> I can implement this, but I can't shake the feeling that something is off in the build configs and that it would be the right thing to fix this before it trips up anyone else in the future.

Perhaps I can assuage your doubts. Here is the relevant excerpt from your Android build the other day. The error occurs when building `monster.cpp.o`.

https://github.com/diasurgical/DevilutionX/actions/runs/19592092869/job/56111698213
```
  FAILED: Source/CMakeFiles/libdevilutionx_monster.dir/monster.cpp.o 
  ...
  In file included from /home/runner/work/DevilutionX/DevilutionX/Source/monster.cpp:42:
  /home/runner/work/DevilutionX/DevilutionX/Source/dvlnet/packet.h:13:10: fatal error: 'sodium.h' file not found
     13 | #include <sodium.h>
        |          ^~~~~~~~~~
  1 error generated.
```

If you check `Source/CMakeLists.txt`, you can see that sodium is linked to the `libdevilutionx` library directly.

https://github.com/diasurgical/DevilutionX/blob/df325f6366f45dc4e37ed3bc355a45cd9d51a52d/Source/CMakeLists.txt#L1003

Meanwhile, `monster.cpp` is a source file for the `libdevilutionx_monster` library.

https://github.com/diasurgical/DevilutionX/blob/df325f6366f45dc4e37ed3bc355a45cd9d51a52d/Source/CMakeLists.txt#L535-L538

And like sodium, `libdevilutionx_monster` is also linked to `libdevilutionx` directly.

https://github.com/diasurgical/DevilutionX/blob/df325f6366f45dc4e37ed3bc355a45cd9d51a52d/Source/CMakeLists.txt#L924

So they are both dependencies of `libdevilutionx`, but sodium is not a dependency of `libdevilutionx_monster`. It would be wrong to try to fix this by adding a dependency where we don't need one. That would only serve to add unnecessary dependencies to our tests that target the `libdevilutionx_monster` library. The fact that it fails to build is more of a feature than a pitfall, since it informs us when we've violated separation of concerns.

File: CMake/finders/Findsodium.cmake [hunk 2]
```diff
@@ -251,47 +251,49 @@ if (WIN32)
...
-else()
-    set(_LIB_TYPE SHARED)
-endif()
-add_library(sodium ${_LIB_TYPE} IMPORTED)
-
-set_target_properties(sodium PROPERTIES
-    INTERFACE_INCLUDE_DIRECTORIES "${sodium_INCLUDE_DIR}"
-    IMPORTED_LINK_INTERFACE_LANGUAGES "C"
-)
+# create imported target only when found, to avoid clashing with in-tree build
```
Comment (AJenbo) [inline CMake/finders/Findsodium.cmake]: Move leaveinfo_t into a header in libdevilutionx instead of modifying the build system.

**Comment by skypher** [comment]
Makes sense. I'll revert the cmake changes and implement the split-out.

**Comment by skypher** [comment]
Alright, I think we're done here. Let me know if this looks good.
